### PR TITLE
👨‍🏭 respect xrefProperties to generate xref spec

### PR DESCRIPTION
**Pull request by OsmondJiang**
#4067

### CODE REVIEW

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 1]
```diff
@@ -35,15 +36,21 @@ public JsonSchemaTransformer(JsonSchema schema)
 
             void TransformXrefScalar(JsonSchema schema, JValue value)
             {
-                extensions[value.Path] = new Lazy<JValue>(
-                    () => TransformScalar(schema, file, context, value, errors, buildChild: null),
-                    LazyThreadSafetyMode.PublicationOnly);
+                var name = value.Path.Split('.').Last();
```
Comment (live1206) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs]: property name could contain `.`, like `ms.date`

Comment (OsmondJiang) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: u r right!

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 2]
```diff
@@ -54,6 +67,7 @@ private void Traverse(JsonSchema schema, JToken token, Action<JsonSchema, JValue
                 case JArray array:
                     if (schema.Items != null)
                     {
+                        schema.Items.Parent = schema.Parent;
```
Comment (mingwli) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs]: still confused about this line

Comment (OsmondJiang) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: go through the json schema, array is different with object

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 3]
```diff
@@ -35,15 +37,26 @@ public JsonSchemaTransformer(JsonSchema schema)
...
+                if (schema.Parent?.XrefProperties.Contains(name) ?? false)
+                {
+                    extensions[value.Path] = new Lazy<JValue>(
+                        () => TransformScalar(schema, file, context, value, errors, buildChild: null),
+                        LazyThreadSafetyMode.PublicationOnly);
+                }
             }
         }
 
         private void Traverse(JsonSchema schema, JToken token, Action<JsonSchema, JValue> transformScalar)
```
Comment (yufeih) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs]: Use a property in a data model with `[JsonIgnore]` to pass parameters is __NOT__ thread safe. The same JsonSchema instance is shared by multiple concurrent calls to Traverse.

```suggestion
        private void Traverse(JsonSchema schema, JsonSchema parentSchema, JToken token, Action<JsonSchema, JsonSchema, JValue> transformScalar)
```

Comment (OsmondJiang) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: OK, but can we make it not shared with multiple concurrent calls just like the JToken passed in?

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 4]
```diff
@@ -66,7 +71,7 @@ private void Traverse(JsonSchema schema, JToken token, Action<JsonSchema, JValue
                     {
                         if (schema.Properties.TryGetValue(key, out var propertySchema))
                         {
-                            Traverse(propertySchema, value, transformScalar);
+                            Traverse(propertySchema, value, transformScalar, schema);
```
Comment (yufeih) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs]: If you pass `key` down to traverse, than the complex `GetPropertyNameFromJsonPath` is not needed
```suggestion
                            Traverse(propertySchema, value, key, transformScalar, schema);
```

Comment (OsmondJiang) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: don't want to do that  since it's change too much of Traverse and GetPropertyNameFromJsonPath is also used by XrefMapBuilder

**Comment by OsmondJiang** [comment]
It's a expected integration test failure, the uid will not be loaded if we can't found corresponding json schema.

image

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 5]
```diff
@@ -22,41 +23,66 @@ public JsonSchemaTransformer(JsonSchema schema)
         public (List<Error> errors, JToken token) TransformContent(Document file, Context context, JToken token, Action<Document> buildChild)
         {
             var errors = new List<Error>();
-            Traverse(_schema, token, (schema, value) => value.Replace(TransformScalar(schema, file, context, value, errors, buildChild)));
+            Traverse(_schema, token, (_, schema, name, value) => value.Replace(TransformScalar(schema, file, context, value, errors, buildChild)));
             return (errors, token);
         }
 
-        public (List<Error> errors, Dictionary<string, Lazy<JValue>>) TransformXref(Document file, Context context, JToken token)
+        public (List<Error> errors, Dictionary<string, (bool, Dictionary<string, Lazy<JValue>>)> properties) TransformXref(Document file, Context context, JToken token)
```
Comment (live1206) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs]:
```suggestion
        public (List<Error> errors, Dictionary<string, (bool isRoot, Dictionary<string, Lazy<JValue> properties>)> xrefProperties) TransformXref(Document file, Context context, JToken token)
```

Comment (live1206) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: shall we follow the naming convention of the return type? it helps me to understand what these mean, especially for this kind of complex types

File: src/Microsoft.Docs.Template/Models/TestData.cs [hunk 6]
```diff
@@ -1,6 +1,7 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using Newtonsoft.Json;
```
Comment (yufeih) [inline src/Microsoft.Docs.Template/Models/TestData.cs]: do we still need to update this test model? This could be removed since we are moving to json schema.

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 7]
```diff
@@ -22,55 +23,135 @@ public JsonSchemaTransformer(JsonSchema schema)
...
+                                xrefPropertiesGroupByUid[uid] = (isRoot, new Dictionary<string, Lazy<JToken>>());
+                            }
+                        }
+
+                        foreach (var (key, value) in obj)
+                        {
+                            if (uid != null && schema.XrefProperties.Contains(key))
+                            {
+                                var propertySchema = TryGetPropertyJsonSchema(schema, key, out var subSchema) ? subSchema : null;
+                                xrefPropertiesGroupByUid[uid].Item2[key] = new Lazy<JToken>(
```
Comment (live1206) [inline src/docfx/lib/schema/JsonSchemaTransformer.cs]: can we do this object implementation in a separate PR?

File: src/Microsoft.Docs.Template/Models/TestData.cs (hunk 6 above)
Comment (OsmondJiang) [reply src/Microsoft.Docs.Template/Models/TestData.cs]: I will removed it in #4694  PR

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 8]
```diff
@@ -22,57 +23,130 @@ public JsonSchemaTransformer(JsonSchema schema)
...
+                node.Replace(TransformScalar(subSchema, file, context, node as JValue, errors));
+
+                return default;
+            });
+        }
+
+        private void Traverse(JsonSchema schema, JToken token, Func<JsonSchema, JToken, string[]> transform)
         {
             schema = _definitions.GetDefinition(schema);
+            if (schema == null)
```
[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: schema for xref properties may be null

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 9]
```diff
@@ -22,57 +23,130 @@ public JsonSchemaTransformer(JsonSchema schema)
...
+            {
+                return false;
+            }
+
+            if (jsonSchema.Properties.TryGetValue(key, out propertySchema))
+            {
+                return true;
+            }
+
+            if (jsonSchema.AdditionalProperties.additionalPropertyJsonSchema != null)
```
[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: AdditionProperties is supported in schema validation, so I'd like to support it in transfrom.

File: src/docfx/lib/schema/JsonSchemaTransformer.cs [hunk 10]
```diff
@@ -22,57 +23,130 @@ public JsonSchemaTransformer(JsonSchema schema)
...
+            Traverse(_schema, token, (schema, node) =>
             {
-                extensions[value.Path] = new Lazy<JValue>(
-                    () => (JValue)TransformScalar(schema, file, context, value, errors), // todo: support JToken
-                    LazyThreadSafetyMode.PublicationOnly);
-            }
+                if (node is JObject obj)
+                {
+                    var uid = obj.TryGetValue("uid", out var uidValue) && uidValue is JValue uidJValue && uidJValue.Value is string uidStr ? uidStr : null;
+                    if (uid != null)
```
[the remark this answers is not in the document]
Comment (OsmondJiang) [reply src/docfx/lib/schema/JsonSchemaTransformer.cs]: good point!
